app/gui2/src/util/ast/reactive.ts

```diff
@@ -0,0 +1,66 @@
+import type { AstId } from 'shared/ast'
+import { type FixedMap, MutableModule } from 'shared/ast'
+import { customRef, markRaw } from 'vue'
+import * as Y from 'yjs'
+
+const WrappedValue = Symbol('wrapped')
+
+export function reactiveModule(doc: Y.Doc): {
+  module: MutableModule
+  triggerAst: (id: AstId) => void
+} {
+  const triggers = new WeakMap<object, () => void>()
+  const tracks = new WeakMap<object, () => void>()
+  const wrap = <Fields>(wrapped: FixedMap<Fields>) => {
+    if (Object.hasOwnProperty.bind(wrapped)(WrappedValue)) return wrapped
+    const track = () => {
+      const track = tracks.get(wrapped)
+      if (track) {
+        track()
+      } else {
+        const { track, trigger } = customRef((track, trigger) => ({
+          get: () => ({ track, trigger }),
+          set: () => {},
+        })).value
+        triggers.set(wrapped, trigger)
+        tracks.set(wrapped, track)
+        track()
+      }
+    }
+    return Object.defineProperty(
+      {
+        get: <Key extends string & keyof Fields>(key: Key) => {
+          track()
+          return wrapped.get(key)
+        },
+        entries: () => {
+          track()
+          return wrapped.entries()
+        },
+        clone: () => {
+          track()
+          return wrapped.clone()
+        },
+        has: (key: string) => {
+          track()
+          return wrapped.has(key)
+        },
+        toJSON: () => {
+          track()
+          return wrapped.toJSON()
+        },
+        set: <Key extends string & keyof Fields>(key: Key, value: Fields[Key]) =>
+          wrapped.set(key, value),
+      },
+      WrappedValue,
+      Object.assign(Object.create(null), { value: wrapped }),
+    ) satisfies FixedMap<Fields>
+  }
+  const module = markRaw(new MutableModule(doc, wrap))
+  const triggerAst = (id: AstId) => {
+    const fields = module.get(id)?.fields
+    if (!fields) return
+    triggers.get((fields as any)[WrappedValue])?.()
+  }
+  return { module, triggerAst }
```

Comment:
I think this is a very good first iteration for this code and I'd be happy to merge it just with minor cleanups. But I think there is an opportunity to offload a lot of this code to already existing Vue's reactive object wrappers, or at least to mimic them a little closer. So treat this whole comment as food for thought (or just as my perfectionism speaking, so you can ignore it completely).

---

With all that out of the way, I think the wrapping code could be simplified a lot by actually utilising a bit more what's already there in vue's reactivity system. Currently it sort of reproduces vue's deps and watchers with this custom `tracks` and `triggers` mapping, but Vue already internally tracks [each field access on each individual reactive proxy](https://github.com/vuejs/core/blob/29e4a1109a8d2a70eec9beda873fde2f4cac5a5b/packages/reactivity/src/reactiveEffect.ts#L13-L18). So we shouldn't need another layer on top. You could probably have just one "master reactive object" per whole module, assuming you only want to track on node-level granularity. Then you could pass the AstId into the [`track`/`trigger`'s key argument](https://github.com/vuejs/core/blob/29e4a1109a8d2a70eec9beda873fde2f4cac5a5b/packages/reactivity/src/reactiveEffect.ts#L33) and that should do the whole trick without any extra maps on our side. We just have to be sure that we perform the cleanup to not grow the deps list forever (i.e. call `trigger` with `TriggerOpTypes.DELETE` when node is removed from module).

---

Alternatively, we can push things a bit and implement fine-grained reactivity for each YJs field. The way I originally imagined it is to either "slightly misuse" or copy and adjust the [`shallowReactive`](https://github.com/vuejs/core/blob/29e4a1109a8d2a70eec9beda873fde2f4cac5a5b/packages/reactivity/src/reactive.ts#L134)'s code path for maps. If we somehow convince the `shallowReactive` function that `Y.Map` is an instance of `Map` (or to be more precise, that it deserves to be classified as `TargetType.COLLECTION`) , then it would already handle `get`, `set` and all the relevant iterators for us. That would actually give us granular tracking. If that isn't sufficient, then we would need our own set of handlers for the proxy. There is of course one remaining source of changes that we need to address - external Y.Map updates. Currently you specialized the update to be ast-node-specific, by using the `triggerAst`. However if we implement the reactivity on on Y.Map fields level, then we could directly use YDoc's `observe` to dispatch appropriate `triggers`. That would mean we don't even need to track changed nodes manually at all.

---

Alternatively, we could simply use an existing "reactive YJs layer" that more or less does the above (with some framework-agnostic reactivity abstraction): https://syncedstore.org/